Add cpp_double_fp and exercise arithmetic_tests

ckormanyos opened the pull request:
The purpose of this pull request is to add the `cpp_double_fp_backend<>` template class.

In the initial commit, we:

- Run the arichmetic tests.
- Modify elementary function tests to prepare for `TEST_CPP_DOUBLE_FLOAT`, but do not run them yet.
- Try some further small adaptions to `arithmetic_tests-hpp` that improve the testing logic for mixed-int128/float128 interaction.

If this runs through, I can add the function tests (stuff like `sin()`, `exp()`, etc.) as a next step. These runj in the branch but are not yet exercised in this initial commit.

ckormanyos commented:
The first CI run (even with only the tiny step of arithmetic tests and modified but not running function tests) has failed due to a rookie typo involving a missing semicolon.

Running again now...

ckormanyos commented:
OK the last CI run passed on GHA but failed a silly typo on performance_test on Circle CI.

The latest push goes a lot farther with an attempt for:
- correcting the above and also
- adding most of functions-and-limits.

ckormanyos commented:
OK.
- GHA is running green with arithmetic tests and most of functions and limits.
- In drone I think I forgot some needed quadmath inclusions in the JAMfile.
- Circle CI failed on specfun_cpp_dec_float, which might be unrelated to the double-double work.

I will press on trying to correct the Jamfile for the drone runs.

After the next cycle, I can collect the know weaknesses - and there are a few, especially with 10-byte `long` `double` mysteries.

And although this is going well, I'd like to resolve any known mysteries (maybe review needed) before pressing on to specfun and maybe more deep conversoin tests.

ckormanyos commented:
OK. I'll jump the CI a bit and think this latest commit might bring this work to a small juncture.

At this juncture, coull you John (@jzmaddock) and Matt (@mborland) please take a cursory glance at the header/tests and give me a sanity check? If you see something waaayyy wrong or out of line, please give me a heads up?

Most of the tests and things are handled by `cpp_double_fp_backend<>` now at the moment.

Known problems include:
1. Round-tripping and banker's rounding (which may be related) and seem to involve the junction between the low/high parts of the _double_-_fp_.
2. Problems with edge-cases of 10-byte double-long-double on platforms having this 10-byte `long` `double` type. For instance, I get non-finite `log()` of double-long-double's `(max)()`, but only for 10-byte long double. I suppose there might be a classic, Chris, dont't you remember, ... something, or a few `L` suffixes, or similar.

In light of this, I'll retreat to the fork a bit and try to look into at least the 10-byte `long` `double` mysteries prior top going in full for spce_fun.

After solving the mysteries, going for spec_fun and setting up some rudimentary docs, this thing will actually be close to ready.

jzmaddock commented on test/test_pow_df.cpp:
```diff
@@ -0,0 +1,867 @@
+///////////////////////////////////////////////////////////////
+//  Copyright 2011 John Maddock.
```
Sorry for not seeing it, but how does this test differ from test_pow.cpp?

ckormanyos replied:
Hi John thanks for looking. The (new) test known as `test_pow_df.cpp` has parsed the original test and thrown away numbers that are so large that the arguments will definitely overflow `cpp_double_fp_backend<double>`.

So instead of having about 650 listed cases and controls, there remain about 650.

jzmaddock commented:
@ckormanyos my apologies for not being able to help out much right now: but a quick scan over the sources shows nothing obviously wrong.

ckormanyos commented:
> a quick scan over the sources shows nothing obviously wrong

Thanks John (@jzmaddock) I'm actually doing alright. I am always amazed how powerful and deep the Multiprecision test suite(s) are. These are wonderful tests.

I've found a few dependencies on `libquadmath.a` in the final drone CI run that I will eliminate in the limits. Other than that, I'm comfortable to simply keep going.

mborland reviewed (commented):
By and large this looks good. I added some minor comments that are applicable in a number of places.

ckormanyos commented on include/boost/multiprecision/cpp_double_fp.hpp:
```diff
@@ -0,0 +1,2050 @@
+template <typename FloatingPointType>
+constexpr void eval_ldexp(cpp_double_fp_backend<FloatingPointType>& result, const cpp_double_fp_backend<FloatingPointType>& a, int v);
+template <typename FloatingPointType>
+#if (defined(_MSC_VER) && (_MSC_VER <= 1900))
+BOOST_MP_CXX14_CONSTEXPR
+#else
+constexpr
+#endif
```
Thanks Matt (@mborland) I'm doing a rather large rework of the numeric_limits for this work and I'll be doing a better job determining what is and is not really `constexpr` moving forward.

ckormanyos replied:
There's one more detail we'll need to handle. MSVC 2015 (also known as VC 14.0) might have difficulties handling C++14: Relaxed constexpr restrictions. Subjectively, I find this to be the case when working with this compiler locally and in CI.

The _only_ compiler I'm using `BOOST_MP_CXX14_CONSTEXPR` for in my current PR drive is, in fact, VC 14.0. So if we drop that compiler, It'll be OK.

It's problem seems to be multiple return statements in a `constexpr` function. But if you really think of it, I'm not really super-100% sure if multiple return statements are allowed in a C++14 `constexpr` function?

So this might prove to be problematic when going C++14 as we move forward.

ckormanyos commented on include/boost/multiprecision/cpp_double_fp.hpp:
```diff
@@ -0,0 +1,2050 @@
+
+         for (;;)
+         {
+            if (static_cast<std::uint_fast8_t>(p_local & static_cast<std::uint32_t>(UINT8_C(1))) != static_cast<std::uint_fast8_t>(UINT8_C(0)))
```
Oh geez, I've gotten in the habit of always deocrating constants with one of the `UINTn_C`-styls macros. When doing this, I cast to the type and then use the macro needed for the width of the constant. I don't know if this makes sense to anyone else, but I use this kind of rule.

ckormanyos commented:
OK. This seems like it's about to go green, which means that the formal correctness of `cpp_double_fp_backend` is nearing completion.

There are, nonetheless, a whole bunch of steps to do.

- We are actually now farther ahead in the Boost.Multiprecision PR than on our Fork. So I will need to merge from Boost's branch "back to" our GSoC fork. I anticipate doing this today or tomorrow.
- There are a bunch of open items needing attention.

These include:
- Banker's rounding is not fully working. Why? Should it work? Why or Why not?
- The selection of `cpp_double_fp_backend::my_value_max()` is unclear. Can we justify this value?
- `eval_exp()` and `eval_log()` need further optimization.
- Optimize (or prove that is already optimized) the routine for division.
- Optimize (or prove that is already optimized) the routine for `eval_sqrt()`.

The only known _problem_ is with banker's rounding. Otherwise, it's just optimize, tests and docs.

The `constexpr`-ification is moving along very well. In fact, large parts of the double-double backend are already now `constexpr`. For this, I have written quite a bit of `<cmath>` functionality in a true `constexpr`-fashion. But there is more work to go. We need to constexpr-ify `::floor()` and maybe even `::log()` and/or `::exp()` for `float`, `double`, `long` `double` and `__float128`. This is a super-cool area to march forward in.

The next step for Boost testing will be to run the `cpp_double_fp_backend` through the gallery of special-function tests, known as the test-suite "specfun". We will probably want to do this locally at first. Also I honestly don't know exactly how to hook up a new Multiprecision backend to the specfun tests.

So I'll ask John (@jzmaddock), very briefly in a few keywords, how do I hook up a new backend to specfun testing in CI? briefly.

ckormanyos commented:
> briefly in a few keywords, how do I hook up a new backend to specfun testing in CI?

Hi @jzmaddock I think I'm seeing the pattern.

I need to:

- Pick up the test files in Jamfile for my particular group.
- Define the setup in the appropriate location(s) in source code.
- Define the test suite in Jamfile. Let's call it`specfun_cpp_double_double`.
- Select my backend of choice.
- And the run a CI job (or command-line job) for the new `specfun_cpp_double_double`-suite.

I'll try in my fork first.
-

jzmaddock commented:
Yup, duplicate this rule: https://github.com/boostorg/multiprecision/blob/8de45037b62b3405139d73d7a14acf0831bf63b6/test/Jamfile.v2#L1472 and rename the tests that are generated to match your new types name, then add the rule invocation below to actually create the testing target like this: https://github.com/boostorg/multiprecision/blob/8de45037b62b3405139d73d7a14acf0831bf63b6/test/Jamfile.v2#L1521

Then modify https://github.com/boostorg/multiprecision/blob/develop/test/math/setup.hpp to set up testing of the new type, again copy the float128 case is the easiest way for what you have.

Finally, duplicate and modify this line: https://github.com/boostorg/multiprecision/blob/8de45037b62b3405139d73d7a14acf0831bf63b6/test/math/instances/Jamfile.v2#L73 to create the template instantiations of the special functions.  This shouldn't be necessary, but it speeds up compilation to do things this way.

And I think/hope that's the lot.

ckormanyos commented:
> Yup, duplicate this rule: ...

Copy that John (@jzmaddock). It looks like the struggle with VC 14.0 continues a tiny bit. And maybe some other lint. But I'm ready to give specfun for `cpp_double_double` a try. I expect some tolerances will fail and some compiler issues may crop up.

I keep looking ahead to the future, but after all that (and I do expect success), then the adaption of _quad_-_float_ (from the original feature-request) should get a lot easier...

Cc: @sinandredemption

ckormanyos commented:
And what does all this mean in the practical sense.

I actually think we have a fair shake (and i'd like to give it a fair try) to get `cpp_double_fp_backend<T>` into 1.82 (with docs, funcs and tests and CI).
Then I think later in the year we can take a more coherent run at _quad_-_float_, maybe even after getting some of the field experience with _double_-_float_.

Cc: @jzmaddock and @sinandredemption and @mborland

jzmaddock commented:
Tolerances should be broadly in agreement with similar backends - so say cpp_bin_float_quad.  The special fun tests are actually very good at really hammering basic arithmetic and have uncovered more than a few bugs in the past, so if you see anything wildly out, I would suspect a bug somewhere.

ckormanyos commented:
> The special fun tests are actually very good at really hammering basic arithmetic and have uncovered more than a few bugs in the past, so if you see anything wildly out, I would suspect a bug somewhere.

Indeed that's for sure John (@jzmaddock). I'm psyched on finding out how it goes. I guess there are still a few remaining problems in CI so I'll work through these.

This is the first time I've tried to add an entire backend on my own. It's challenging but a great learning experience.

Also to be clear, I'm only (we're only) going for `cpp_double_fp_backend` in this PR and (maybe) for 1.82. The quead stuff will come later.

Cc: @sinandredemption

ckormanyos commented:
This first step took a while and has now _gone_ _green_.
But we are **not yet** ready to merge.

We still need the following:
1. Test and pass specfun (might take a while)
2. Write and build small docs on `cpp_double_fp_backend<T>`.
3. Review.

I am happy with this first preliminary green milestone and will now turn to points 1 and 2, beginning with specfun.

At first I'll prepare this in our GSoC fork.

Cc: @jzmaddock and @sinandredemption and @cosurgi and @mborland

ckormanyos commented:
Hi John (@jzmaddock),

I wonder if you could help just a tiny bit with guidance? Much to my delight my local tests of `specfun` running on GCC-9, -std=c++17 were remarkably successful on a first run. I received the following last lines of a build log...

I got `...updated 337 targets...` but also `...failed updating 12 targets...` (see second text block below).

Now I need to see the failures.

From these failures on `Zeta`,  do you have any guidance on how I can actually run the relevant test case and put it under a microscopic debug session? I feel like I do not know what the row or mean are all about? See first text block below:

For `Zeta`

```sh
====== BEGIN OUTPUT ======
Running 1 test case...
Tests run with GNU C++ version 9.3.0, GNU libstdc++ version 20200808, linux
Zeta: Random values greater than 1 with type cpp_double_double
zeta<cpp_double_double> Max = 0.1958 RMS Mean=0.0706
    worst case at row: 12
    { 4.988127231597900390625, 1.037268683773212337925891820825986369163 }


Zeta: Random values less than 1 with type cpp_double_double
zeta<cpp_double_double> Max = 147.7 RMS Mean=13.23
    worst case at row: 69
    { -17.99382781982421875, -0.08425805633923885570999830270532256204479 }


Zeta: Values close to and greater than 1 with type cpp_double_double
zeta<cpp_double_double> Max = 0.19 RMS Mean=0.08633
    worst case at row: 22
    { 1.49527740478515625, 2.631125733364186076058384433696635480314 }


Zeta: Values close to and less than 1 with type cpp_double_double
zeta<cpp_double_double> Max = 1.003e+04 RMS Mean=2735
    worst case at row: 16
    { 0.999987542629241943359375, -80273.18355079243243774478653269383556504 }
Peak error greater than expected value of 1000
../../../libs/math/test/handle_test_result.hpp(165): [1;31;49merror: in "test_main": check bounds.first >= max_error_found has failed[0;39;49m
Mean error greater than expected value of 200
../../../libs/math/test/handle_test_result.hpp(170): [1;31;49merror: in "test_main": check bounds.second >= mean_error_found has failed[0;39;49m


Zeta: Integer arguments with type cpp_double_double
zeta<cpp_double_double> Max = 8.453 RMS Mean=2.34
    worst case at row: 54
    { 56.0, 1.0000000000000000138777878097252328 }



[1;31;49m*** 2 failures are detected in the test module "Master Test Suite"
[0;39;49m
EXIT STATUS: 201
====== END OUTPUT ======
```

Build result:

```sh
...failed updating 12 targets...
...skipped 12 targets...
...updated 337 targets...
```

Now I'm having difficulty identifying exactly which lines/cases/tests are needing special attention.

jzmaddock commented:
So if we look at:

```
Zeta: Values close to and less than 1 with type cpp_double_double
zeta<cpp_double_double> Max = 1.003e+04 RMS Mean=2735
    worst case at row: 16
    { 0.999987542629241943359375, -80273.18355079243243774478653269383556504 }
Peak error greater than expected value of 1000
```

The case causing the failure is `zeta(0.999987542629241943359375)` which should yield -80273.18355079243243774478653269383556504 but is 10^4eps out compared to an expected tolerance of 10^3.  Looking at the expected results for other types, this does indeed look like a _slight_ outlier.

That case should go through:

```
template <class T, class Policy>
T zeta_imp_prec(T s, T sc, const Policy&, const std::integral_constant<int, 113>&)
```

which is a rational approximation and should be super accurate - maybe check that the constant coefficients are being initialized correctly?  For comparison, maybe check the error rate for cpp_bin_float_quad.

BTW I've been meaning to create a "dual precision" backend for multiprecision that would let you run the same calculation on two backends simultaneously and see where they they start to differ.

ckormanyos commented:
> Just tried:
> get an error of 0, which looks good to me, so the difference with the double double looks like it might be real?

Indeed. I stepped through the impl. Your exact result makes sense to me. It especially makes sense, now that I study the implementation of `zeta()` in that region) which is solely based on algebraic polynomial expansion.

I fear this might take us all the way back to some fundamental bit-losses in the basic operations add/sub/mul/div in the low-level implementation of `cpp_double_fp_backend<>`.

We have an [open issue](https://github.com/BoostGSoC21/multiprecision/issues/35) (one of the first issues from July 2021) in the GSoC about this.

I fear that this type (its low-level eval_ops might not be _fully_ _cooked_ yet.

ckormanyos commented:
> get an error of 0, which looks good to me

Ahhh. OL, John (@jzmaddock), well, I figured out at least _partly_ what's going on, but I don't know the best way to fix it entirely in the backend/test/etc. world.

- The input argument is an exact `double`. However in the initialization is carried out with string-conversion, since we are above 64 digits, yet below 113 digits in the type.
- This causes imprecision in the digits immediately following the `...375` in the input.
- From that point on, the calculatoin carries this fuzz through.

If I simple use the `double` representation of the number to construct the argument `x` to `zeta(x)`, then the input and following calculation retain exactness.

I'll need to see if the backend (on its own) can handle the case decision to initialize with the native floating-point type of string. I think I might need more constructors. Or if the backend can't handle this alone, then we need to think harder.

jzmaddock commented:
Is that not an error in string conversion then: I would assume it should convert to an exact binary value?

ckormanyos commented:
>> Is that not an error in string conversion then: I would assume it should convert to an exact binary value?

> hope to find a solution

Hi John (@jzmaddock) regarding string input for `cpp_double_fp_backend`, after a while of thinking it's clear that you already solved this problem with `cpp_bin_float<>`. I racked my brain and found that for a binary backend, that's the working model. We need to (basically just like a compiler) support another multiprecision to fully interpret multiprecision strings.

This will pull `cpp_int` into the game. But I'm OK with that.

Along the lines of getting it working and then optimizing it, I find a usage of the `cpp_bin_float`-way for string-input is what is needed. Anything 3else does not capture the digits.

I do not know if my ramblings make sense. But If so, I'd duplicate your code, test it, then with you together, find a way to make it _publicly_ available for binary backends.

Thoughts? Am I crazy?

jzmaddock commented:
> Thoughts? Am I crazy?

Probably crazy to want to take on that much coding, but yes, I fear that is the correct way to go.  The hard part is probably packing the bits in the resultant integers into the float type.

sinandredemption commented:
> Probably crazy to want to take on that much coding

@ckormanyos Please do let me know in chats if you need an extra hand. I've been trying to start contributing again.

cosurgi commented:
It is great to see you going forward with this. Unfortunately I cannot join right now: too much work at the university at the moment :( But I will keep monitoring your changes, hoping to jump in when it becomes possible.

ckormanyos commented:
>> you already solved this problem with cpp_bin_float<>

> want to take on that much coding, but yes, I fear that is the correct way to go

Thanks for your input John. Yes. I will try - based on Your code that is already fully tested, well-established and cleanly available in `cpp_bin_float`.

This will have a bit of coding depth to it, so best to hammer this particular detail in our fork.

ckormanyos commented:
> let me know in chats if you need an extra hand.

Hi Fahad (@sinandredemption) we'll be moving back and forth to/from our GSoC fork for a couple of the more in-depth corrections.

This PR has already shown that:

- `cpp_double_fp_backend` seems robust enough to become a reliable badkend.
- We still need to work on exact string read.
- I'd like to study if we do/do-not have optimal algorithms for add/sub/mul/div/sqrt.
- Passing all of `specfun` still needs a bunch of work locally like in our fork.

I am going to start the first ponit back in our fork. I'll be posting more details in the issues of the fork and we can keep this PR open as we progress.
-

ckormanyos commented:
Well, this one is getting close, but there still seems to be a slight problem with rounding discovered in association with `modf`.

I think I need some advice. I'm losing 1-eps when computing and recombining the integral and fractional parts.

- Is there any reason _why_ one might believe that integral/fractional calculation and recombination might lose 1-eps?
- Or might it rather be, as I sadly suspect, that there could be an underlying 1-eps problem in one or some of add/sub/mul/div?
- Or is the double-float type lossy by nature? I thought (naively) it was as precise as its underlying components.

I put a [workaround](https://github.com/boostorg/multiprecision/blob/f6ca99d93a372b48098e07a7ffa06601f8c2f4b5/test/test_round.cpp#L185) in `test_round.cpp`. For the `cpp_double_fp_backend` class there is a problem when computing and recombining the integral and fractional parts of floating-point. It is about $1/1000$ or so, so a kind of edge case. But I have not found the pattern when a candidate needs 1-eps.

All the other types being tested have equality on the sum of the fractional part and integral part being equal to the original number. But I need `epsilon()` for the `cpp_double_fp_backend` case. Every other test in the Multiprecision suite (even specfun) is now going green.

Cc: @jzmaddock and @cosurgi and @sinandredemption and @mborland

cosurgi commented:
I hope that this will fix this last problem. Thank you so much for working on this.

ckormanyos commented:
So folks, I have a few more issues in covrage and documentation ongoing. But the development drive on `cpp_double_fp_backend` has for the most part concluded.

The time for reflection and reviewing is upon us.

I will be moving this forward for Boost soon. I'm not sure if 1.89 is the right call. I think 1.90 is a better fit. But if we get agreement soon, we can push for 1.89.

So please if you get a chance, take a look at this work, review it critically. Contribute comments or doubts.

Cc: @jzmaddock and @sinandredemption and @cosurgi and @mborland

mborland commented:
Here is my local output from test round since I can replicate the error:

```
...
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
Error at this spot hint: what is going on here?
test_round.cpp(97): Rounded result differed by more than 0.5 from the original in function 'typename std::enable_if<!boost::multiprecision::is_interval_number<T>::value>::type check_within_half(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long]'
Values were: 9223372036854775807                     9223372036854775806
test_round.cpp(153): Rounded result differed by more than 1 from the original in function 'void check_trunc_result(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long]'
Values were: 9223372036854775807                     9223372036854775806
test_round.cpp(97): Rounded result differed by more than 0.5 from the original in function 'typename std::enable_if<!boost::multiprecision::is_interval_number<T>::value>::type check_within_half(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long]'
Values were: 9223372036854775806                     9223372036854775805
test_round.cpp(153): Rounded result differed by more than 1 from the original in function 'void check_trunc_result(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long]'
Values were: 9223372036854775806                     9223372036854775805
test_round.cpp(97): Rounded result differed by more than 0.5 from the original in function 'typename std::enable_if<!boost::multiprecision::is_interval_number<T>::value>::type check_within_half(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long long]'
Values were: 9223372036854775807                     9223372036854775806
test_round.cpp(153): Rounded result differed by more than 1 from the original in function 'void check_trunc_result(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long long]'
Values were: 9223372036854775807                     9223372036854775806
test_round.cpp(97): Rounded result differed by more than 0.5 from the original in function 'typename std::enable_if<!boost::multiprecision::is_interval_number<T>::value>::type check_within_half(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long long]'
Values were: 9223372036854775806                     9223372036854775805
test_round.cpp(153): Rounded result differed by more than 1 from the original in function 'void check_trunc_result(T, U) [T = boost::multiprecision::number<boost::multiprecision::backends::cpp_double_fp_backend<long double>, boost::multiprecision::et_off>, U = long long]'
Values were: 9223372036854775806                     9223372036854775805
16 errors detected.

EXIT STATUS: 16
====== END OUTPUT ======
```

mborland commented:
Ubuntu-jammy-specfun seems fine on Github Actions now. I added linking of quad math to your rule defining tests. Still trying to figure out what's up with rounding on Mac.

ckormanyos commented on include/boost/multiprecision/cpp_df_qf/cpp_df_qf_detail_ccmath_frexp.hpp:
```diff
@@ -0,0 +1,38 @@
+   // Default to the regular frexp function.
+   using std::frexp;
+
+   return frexp(arg, expptr);
```
Thanks Matt,

I'm aware of this. I need to do something akin to what you did in Boost.Math's `ccmath` work.

But for now, I am going to kick these (and all of the similar ones) down the road a bit. I need to get the formal testing and correctness done first.

I'd like to pick up this discussion later. I don't know if you ever did or didn't get all this to work out in `ccmath` yet?

mborland replied:
Everything in ccmath should be working for the C++23 coverage. I haven't done the C++26 portions like log yet. Those got merged into the draft standard recently.

ckormanyos commented on include/boost/multiprecision/cpp_df_qf/cpp_df_qf_detail.hpp:
```diff
@@ -0,0 +1,169 @@
+#ifndef BOOST_MP_CPP_DF_QF_DETAIL_2023_01_02_HPP
+#define BOOST_MP_CPP_DF_QF_DETAIL_2023_01_02_HPP
+
+#include <boost/config.hpp>
```
I will need a tad of help on this one, Matt. But that's also down the road a few days.

ckormanyos commented:
>  Still trying to figure out what's up with rounding on Mac.

There are 5 or six versions of `eval_convert_to` and I can see how _ALL_ of them need to be reworked now based on your input.

Thanks, many thanks Matt (@mborland).

I have reworded the 1 `eval_convert_to` that seems to be causing grief and I'd like to see how it shakes out. Then I will rework all of them accordingly.

ckormanyos commented:
> If you copy and paste off of develop

Copy and paste _what_ off of develop? I can't see any big diffs in our YAMLs? Or is it picking up runners from elsewhere?

mborland commented:
> > > If you copy and paste off of develop
> > 
> > 
> > Copy and paste _what_ off of develop? I can't see any big diffs in our YAMLs? Or is it picking up runners from elsewhere?
> 
> Sorry I just looked and the multiprecision.yml hasn't been updated. I did math last week and thought I did this one too. I'll make the updates.

CI will be fixed here: https://github.com/boostorg/multiprecision/pull/700

ckormanyos commented:
Thanks for all this Matt.

I did a bit more cleanup too.

But I need to see what's going on with the standalone tests. I won't get to this until tomorrow though.

mborland commented:
I applied the same Apple workaround to ARM64 and it seems to do the trick. There was a missing `constexpr` on the new `extract` function which caused a number of failures. Fingers crossed this should run Green

ckormanyos commented:
> I applied the same Apple workaround to ARM64 and it seems to do the trick.

Oh wow. Yeah that is really nice how you did that. Thank you Matt.

> There was a missing `constexpr` on the new `extract` function which caused a number of failures. Fingers crossed this should run Green.

Uggghhh how could I forget that. Well I _could_ do some more fiddling around with some of the `eval_convert_to` functions. But I'm going to stop fiddling around soon and leave all that to the evolution of this thing.

Let's see if we can  get green. Then I think it's time to start discussing `develop`.

Cc: @jzmaddock and @cosurgi and @sinandredemption

ckormanyos commented:
> Looks like this is green across the board excluding the known to be broken inspect. Coverage at 98.2% of the diff for the PR is impressive given the size.

Yes.

I need to do one final action. I think I need to merge all this _BACK_ to the original GSoC repo. Then verify the merge manually with a merge tool.

I would like the PR to come from the GSoC21 repo. Only then can the change log and commit history be preserved.

I think I can handle this tomorrow. After that, it's full speed ahead on Double FP.

- Thanks Fahad (@sinandredemption) and Janek (@cosurgi) for the amazing work in the GSoC and your help beyond that.
- Thank you Matt (@mborland). Just a week working with you cleared up numerous issues.
- And thanks John (@jzmaddock) for your guidance and patiently helping us and guiding us to this point.

Sure this library part will evolve. But it's time to go to user-land with it.

I'll do the proper merging and PR this from GSoC 21. I can handle this alone (I hope).

ckormanyos commented:
This PR did not preserve the full commit log from thee original GSoC project. So I have replaced this one with #701 which corrects this. If you have been following this PR, it is replaced by #710.

Closing.

ckormanyos commented:
> I has been a pleasure working with you.

You to Janek. I think this thing will evolve a bit when it goes live. So my plan is to see how that shapes up for a while, what we learn.

And after a learning period, we could go after `cpp_quad_fp_backend`, which will have its own set of challenges, but be a lot easier after learning so much from the double-word case.
